code/code/misc/info.cc

```diff
@@ -5385,642 +5385,291 @@ void TBeing::describeTrapDamType(const TTrap *obj, int) const
+    return std::find(validSpellTypes.begin(), validSpellTypes.end(), spellType) != validSpellTypes.end();
   };
 
-  if (!hasClass(CLASS_SHAMAN) && !isImmortal()) {
-    sendTo("You aren't a Shaman, therefore you have no use for rituals.\n\r");
-    sendTo("Perhaps using the SPELLS command is what you want.\n\r");
-    return;
-  }
+  static const std::map<skillUseClassT, unsigned short> classForSpellType = {
+    {SPELL_CLERIC, CLASS_CLERIC},
+    {SPELL_DEIKHAN, CLASS_DEIKHAN},
+    {SPELL_MAGE, CLASS_MAGE},
+    {SPELL_SHAMAN, CLASS_SHAMAN}
+  };
 
-  if (!(d = desc))
-    return;
+  auto countValidClasses = [this]() {
+    int count = 0;
+    for (auto spellType : validSpellTypes) {
+      if (hasClass(classForSpellType.at(spellType)))
+        count++;
+    }
+    return count;
+  };
 
-  *buffer = '\0';
+  std::map<skillUseClassT, std::map<spellCategories, std::vector<spellNumT>>> spellInfoByType;
+  for (auto spellType : validSpellTypes)
+    spellInfoByType.emplace(spellType, createSpellCategoriesMap());
 
-  if (argument.empty())
-    memset(types, 1, sizeof(int) * 4);      
-  else {
-    memset(types, 0, sizeof(int) * 4);
-
-    arg=argument.word(0);
-    arg2=argument.word(2);
-    arg3=argument.word(3);
-
-    if (is_abbrev(arg3, "all"))
-      showall = 1;
-
-    if (!arg2.empty()) {
-      if (is_abbrev(arg2, "all"))
-        showall = 1;
-      else if (is_abbrev(arg2, "targeted"))
-        subtype = 1;
-      else if (is_abbrev(arg2, "nontargeted"))
-        subtype = 2;
-      else
-        badtype = 1;
-    }
-    
-    if (is_abbrev(arg, "offensive")) {
-      if(!subtype || subtype == 1)
-        types[0] = 1;
-      if(!subtype || subtype == 2)
-        types[1] = 1;
-    } else if (is_abbrev(arg, "utility")) {
-      if(!subtype || subtype == 1)
-        types[2] = 1;
-      if(!subtype || subtype == 2)
-        types[3] = 1;      
-    } else
-      badtype = 1;
-    
-    if (badtype) {
-      sendTo("You must specify a valid ritual type.\n\r");
-      sendTo("Syntax: rituals <offensive|utility> <targeted|nontargeted> <all>.\n\r");
-      return;
-    }
-  }
+  std::vector<skillSorter> skillsSorted;
+  skillsSorted.reserve(230);
+
+  for (spellNumT i = MIN_SPELL; i < MAX_SKILL; i++) {
+    spellInfo *spell = discArray[i];
 
-  std::vector<skillSorter>skillSortVec(0);
+    if (!spell)
+      continue;
 
-  for (i = MIN_SPELL; i < MAX_SKILL; i++) {
-    if (hideThisSpell(i) || !discArray[i]->minLifeforce)
+    if (hideThisSpell(i) || !isValidSpellType(spell->typ) ||
+        getDisciplineNumber(i, false) == DISC_NONE ||
+        (!spell->minMana && !spell->minLifeforce && !spell->minPiety))
       continue;
 
-    skillSortVec.push_back(skillSorter(this, i));
+    skillsSorted.push_back(skillSorter(this, i));
   }
-  
-  // sort it into proper order
-  sort(skillSortVec.begin(), skillSortVec.end(), skillSorter());
+  sort(skillsSorted.begin(), skillsSorted.end(), skillSorter());
```

Comment:
So I actually had it as a static before I realized that the `skillSorter` class constructor requires `this` as a parameter because the sort order is based on the character's learnedness of the skills, primarily. So this actually does need to be recalculated every time, unless we want to sort the skills differently.

However, in the newer refactoring I updated this to use `emplace_back` instead of `push_back`, if that matters.